WcaOnRails/spec/controllers/competitions_controller_spec.rb

```diff
@@ -221,6 +221,23 @@
         sign_in organizer
       end
 
+      it 'cannot pass a non-delegate as delegate' do
+        delegate_ids_old = competition.delegate_ids
+        fake_delegate = FactoryGirl.create(:user)
+        post :update, id: competition, competition: { delegate_ids: fake_delegate.id }
+        invalid_competition = assigns(:competition)
+        expect(invalid_competition.errors.messages[:delegate_ids]).to eq ["You must pass only delegates here"]
+        competition.reload
+        expect(competition.delegate_ids).to eq delegate_ids_old
+      end
+
+      it 'can change the delegate' do
+        new_delegate = FactoryGirl.create(:delegate)
+        post :update, id: competition, competition: { delegate_ids: new_delegate.id }
+        competition.reload
+        expect(competition.delegate_ids).to eq new_delegate.id.to_s
```

Comment:
Yeah, I'm not pleased with how I did this delegates thing. If you know how
to clean it up, I'd really appreciate a lesson!

On Tue, Jan 5, 2016 at 3:13 AM, Tim Habermaas <notifications@github.com>
wrote:

> In WcaOnRails/spec/controllers/competitions_controller_spec.rb
> <https://github.com/cubing/worldcubeassociation.org/pull/298#discussion_r48834696>
> :
>
> > @@ -221,6 +221,23 @@
> >          sign_in organizer
> >        end
> >
> > +      it 'cannot pass a non-delegate as delegate' do
> > +        delegate_ids_old = competition.delegate_ids
> > +        fake_delegate = FactoryGirl.create(:user)
> > +        post :update, id: competition, competition: { delegate_ids: fake_delegate.id }
> > +        invalid_competition = assigns(:competition)
> > +        expect(invalid_competition.errors.messages[:delegate_ids]).to eq ["You must pass only delegates here"]
> > +        competition.reload
> > +        expect(competition.delegate_ids).to eq delegate_ids_old
> > +      end
> > +
> > +      it 'can change the delegate' do
> > +        new_delegate = FactoryGirl.create(:delegate)
> > +        post :update, id: competition, competition: { delegate_ids: new_delegate.id }
> > +        competition.reload
> > +        expect(competition.delegate_ids).to eq new_delegate.id.to_s
>
> Looks like we overwrite delegate_ids:
> https://github.com/cubing/worldcubeassociation.org/blob/master/WcaOnRails/app/models/competition.rb#L118
> ActiveRecord's has_many already adds such a method which returns an
> array, that's what confused me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cubing/worldcubeassociation.org/pull/298/files#r48834696>
> .
>